### ISSUE TITLE
Add TS101 support

### ISSUE THREAD

**Issue by rozwell** [Ralim/IronOS#1420]
Hopefully TS101 hardware will be easy to adapt and there won't be any bootloader issues.

**Comment by Ralim** [comment]
Trying to source one. Miniware so far have not shared any details (firmware, schematics).
May or may not be worth my time to to depending on complexity, but will try.

**Comment by River-b** [comment]
> Hopefully TS101 hardware will be easy to adapt and there won't be any bootloader issues.

see if the miniware community is willing  to  either mail Ralim a TS101  or easier if they can put some funds into the Sponsor link in top right of the screen.

Miniware could not be bothered to send Ralim one.  which they should since  his firmware  runs on most all their products and makes people actually like using the Miniware products more.  without IronOS the miniware firmware is not as fun or good.

I know Ralim does this for the love of  IronOS community.  however, the community doesn't send him love and coffee, (but many issues). 

See reference below, fact is there over  155,000  downloads of  open IronOS  _**just for TS100 alone**_   (yet  maybe ~24 Kofi/coffees donated to for all his volunteer work and most of those are from Pinecil community which is small % of downloads).  

If even 10% of people donated  0.25 cents it would be meaningful since Ralim has a day job to pay for rent/food.  I'd like to get that word out. And he tries to do everything he can  to only use features on Github  that don't trigger the Advertisements.

this is not pressure, just putting out some facts to help people know how much work goes into firmware, and how little the community is aware of  it, and ralim never asks,  the heart sponsor link is barely visible.   I also knows he buys a lot of equipment and chargers with his own money to test them and improve issues.

[Reference datasheet](https://docs.google.com/spreadsheets/d/e/2PACX-1vTVXQfzKYp0MsOMG7gf9u-V8gYyewQVI84YdE8i4e2ZBf9xSU--9BgFSd5JIUkVaGwjkzyVK0yPOmvt/pubhtml#)

**Comment by rozwell** [comment]
@Ralim I'm actually surprised they IronOS isn't the default firmware for Miniware irons, since it's just fantastic compared to default. Were you ever in contact Miniware about it? Or them supporting the project in any way? The least they could do, is provide 1 unit...

@River-b good point. I really love IronOS (using on TS100) so I just donated 4 Koffis 😉 
It's good you pointed it out, cause I didn't actually notice that option it until now 😅 - dealing with all the daily crap can make people so dull...

**Comment by rozwell** [comment]
@Ralim, @River-b Miniware responded to my YouTube comment:
> MINIWARE
> [9 hours ago](https://www.youtube.com/watch?v=e7saVBTAgEg&lc=UgweHzJqPc3SchHRdrJ4AaABAg.9gqHlzYzik09guPkiwPXbR)
> Already send to Ralim.

@Ralim Can you confirm? 😀

**Comment by sorgelig** [comment]
I've just got TS101. That's great improvement over TS100. And it's really useful updates unlike Pinecil64 v2.0 (BT addition is dumbest feature ever). I like new larger screen and clean interface. PD support. New slot for tip not requiring to screw it every time you change the tip. The only problem with original FW is bad work with non-original tips with noisy temperature sensors. Otherwise FW is great.

**Comment by Ralim** [comment]
Alrighty my TS101 has arrived.

First a word of warning, don't rush to open it, (1) I broke plastic clips for the back cover (there is no screw). (2) You have to remove the front glass to get the screen out (if you just pull up on the PCB, you will tear off the screen). As such as I have not done a full teardown yet.



First impression notes:

- Same STM32F103CU as the TS80P
- Screen is a tiny bit larger, looks to be same resolution as others
- Their bootloader is 'uge at 32k as they have packed storage into its region instead.
- - This means to make IronOS work either need to replace bootloader, or get creative at moving as much as we can into that area somehow.
- Features are basically same as the old PinecilV1 though less powerful CPU as still on the F1.
- Tip retention with the spring loaded ball, though nothing I would write home about, tips spin when using non-round ones.

Biggest issue is that at the moment only have ~32k of flash to work with, which is _small_.
So will need to ponder this a bit about how best to handle this; will add some complexity to bring up.

Have dumped the firmware off mine, and started poking around, need to figure out pinmaps and do all the usual bringup mess.
Once I confirm OLED+Buttons I will try and get my DFU bootloader working so we have a backup plan.

**Comment by sandmanRO** [comment]
I personally don't like the new design. It's somehow bulky / non-fluid as opposed to the old TS100 design, more likely dictated by the parts size fitted in rather than the other way around. It's very unlikely I would buy this product. But then again, I already have four TS100 units so I would not be needing more portable irons any time soon.
P.S. Please don't shoot...it's just a personal opinion ;-)

**Comment by SubZer01** [comment]
> I personally don't like the new design. It's somehow bulky / non-fluid as opposed to the old TS100 design, more likely dictated by the parts size fitted in rather than the other way around. It's very unlikely I would buy this product. But then again, I already have four TS100 units so I would not be needing more portable irons any time soon. P.S. Please don't shoot...it's just a personal opinion ;-)

and who is interested in your opinion?
it is about the firmware Support from the TS101.
furthermore, the new design is made softer rather than vice versa.

**Comment by discip** [comment]
@SubZer01
>and who is interested in your opinion?

I'm pretty sure you wouldn't be "talking" like that if you knew that this man contributed a significant part of the code that makes your device heat so precisely.

Look here: https://github.com/Ralim/IronOS/issues/1038

> P.S. Please don't shoot...it's just a personal opinion ;-)

Furthermore he literally asked for not being shoot. 😞

**Comment by Ralim** [comment]
@SubZer01 Please watch your tone, you are coming across super rude. I would rather not block anyone, but not against it.

Making some gradual progress on this. Dont have a huge amount of time at the moment, but have started poking a flash dump in Ghidra.

My rough plan is to:
- Figure out the pinout they are using
- - Mildly concerned as they are bit-banging everything like before so wondering how many signals are on hardware pins _sigh_
- Get all the basic features working
- Trace out USB-PD stuff and see how thats done
- Figure out how to use the screen (it is 128x32 so higher resolution)
- - I knew this day was inevitable but have to think about best way to handle it. Worst case will just scale possibly for now (unsure; would love ideas)
- Wonder what to do about bootloaders, their one is super wasteful. They are clearly storing the virtual disk in the bootloader area. Which means we can store settings there but not the bootup logo.


@sandmanRO 
I'm fairly similar tbh. Also dont like that tip contacts are soldered to the board; and assembly of screen under glued glass makes it muuch harder for repair. (Thus why I'm disassembling code rather than multimeter probing it).

**Comment by BF3000** [comment]
> I've just got TS101. That's great improvement over TS100. And it's really useful updates unlike Pinecil64 v2.0 (BT addition is dumbest feature ever). I like new larger screen and clean interface. PD support. New slot for tip not requiring to screw it every time you change the tip. The only problem with original FW is bad work with non-original tips with noisy temperature sensors. Otherwise FW is great.

Actually  Pinecil v2 have much more improvements than TS101. BT support is dumb, but 28v PD support & short tips (3 sec to heat up from 20C to 300C) is not.

**Comment by PUF52** [comment]
Just mini ware posted the firmware 1.09

**Comment by PUF52** [comment]
http://www.minidso.com/forum.php?mod=viewthread&tid=4937
http://www.minidso.com/forum.php?mod=viewthread&tid=4937
firmware is here

**Comment by PUF52** [comment]
really called it 1.10

**Comment by River-b** [comment]
> Actually Pinecil v2 have much more improvements than TS101. BT support is dumb, but 28v PD support & short tips (3 sec to heat up from 20C to 300C) is not.

- the BL706 chip came with BT support as a side bonus and it was low cost and readily available chip unlike other chips  that are in short supply in the world.  This is one main reason the BL706 MCU was chosen for Pinecil V2. 
- Pinecil also serves  the Pine64 FOSS community as a RISC-V  development tool. It is _not just_  a soldering iron like other soldering irons. 
- Some people buy Pinecil and do not have any tip installed. They just have a USB-C  break-out board connected to the back side which helps with development and working with USB-C  libraries for example. They also have a completely different development OS running on it that is not IronOS.
- People often do not realize  that Pinecil is not meant to be just a soldering iron and then criticise these other features it has based on on idea that a soldering irons can only function as a soldering iron. If you think of it _first_  as a RISC-V  development hardware for the open source community  and second as a soldering iron  then it will make more sense why it has the  extra features for open community  to explore and have fun with. 
- See[ Pinecil Wiki here ](https://wiki.pine64.org/wiki/Pinecil) which explains how it serves as a RISC-V development tool 
- Other irons only care if they can solder.  The Pinecil is held to an additional standard - how well does it meet the needs of the open source community purposes to be a RISC-V  dev tool and could features be added that exploit some community requests for fun and science.

**Comment by Ralim** [comment]
I am planning to continue work on this, motivation is lower as I'm going to have to rebuild a lot of the firmware to support this iron, so initial firmware will probably not be great.

So far all of the hardware pinouts I've found were done by an idiot, and so not going to be able to use hardware peripherals for things.
About halfway through making a pi out for the device.


Question for the community: the current bootloader wastes a bunch of space on the device. Is it worth pushing people to replace it or not?

**Comment by sorgelig** [comment]
If flashing bootloader will go wrong, will it brick the device? If so, then it's better to avoid flashing bootloader unless it's absolutely required.

**Comment by Ralim** [comment]
Yeah it can definitely go wrong. Its a problem that can be decided later on I guess. Just concerned since bringing in a new device also implies full feature completeness (langues etc) which is currently a fair bit chunky.

**Comment by sorgelig** [comment]
will it require to flash original bootloader if for some reason an original firmware will be needed?

**Comment by BF3000** [comment]
As long as there will be a way to restore it to original, I'm fine with that.
Flashing to IronOS is usually one way road.
And BIG WARNING will be nice.

**Comment by Ralim** [comment]
At this point I do not know if I will bother; it would most likely be a sub-par experience due to their hardware design decisions.
Most likely answer is that I will add some basic support; but (a) it will take  time and (b) its a lower priority as I'm not paid to waste 100's of hours working around their design decisions and decompiling their firmware.

**Comment by L1cardo** [comment]
They released v2 of their firmware which can go to 240w with PD3.1, that is crazy!!

**Comment by River-b** [comment]
> They released v2 of their firmware which can go to 240w with PD3.1, that is crazy!!
 could you give me a link to where they say it will do  240W? 

 I think they just enabled  PD3.1  like we had in IronOS already for >4 months for Pinecil V2 (also supports EPR PD3.1).   I've been running Pinecil V2 for 4 months with IRonOS  at 28V, but with special Pine Short tip resistence the highest max would be  126W.

Miniware TS101  says  up to 28V  90W now  on their website (the 28V is  PD3.1 current chargers that are available).   There are not even any  USB-C  chargers out yet that do 240w, the highest USB-C chargers that are EPR PD3.1  on the market in 2022  can only do  28V 140W. 
 
Miniware  website says  only 90W, 28V  on USB-C. 
![image](https://user-images.githubusercontent.com/97197236/206842790-e800cdd5-0ae4-4f8d-95a3-f8334b88d0b8.png)

Miniware uses long 7.8-8.3 ohm tips,  so at  28V EPR charger, with some manufacturer tolerance that would be about the 90W they say on their new listing.   But not 240W (since that doesn't exist yet).   A Special  USB-C cable is required.  the cable  is  rated at 240W (future proof),  but the Chargers on the market are max  28V-140w right now.   I know because I've been using it in the  Pinecil V2 iron for  4 months now  a  140W-28V-5Amp charger with  special  USB-c 240w-48v  cable .   Cable  is designed to last  into new  new standards for EPR that will be coming  in the next few years  ,but the chargers are only  capable of the  lower end of the PD3.1

**Comment by Ralim** [comment]
I had to poke them a bunch and give them a hand to get them to add EPR support to the unit.
I'm glad that they did it; but no idea why they didnt at launch.

**Comment by sorgelig** [comment]
I more wonder why 28V can be used on USB-C input but not on DC input. Is there are difference?

**Comment by PauliusMorku** [comment]
> I more wonder why 28V can be used on USB-C input but not on DC input. Is there are difference?

I was wondering the same thing, it is so annoying that we can't use a fully charged 6s battery 😑 or maybe we can? Is there anything on the DC path that could not handle 25.2 volts?

**Comment by L1cardo** [comment]
@Ralim Just got my new ts101 iron, really like the tip slot. Looking forward to testing your new FW, even I should flash a bootloader.
Let me know if you need any of ny help.

**Comment by L1cardo** [comment]
> > I more wonder why 28V can be used on USB-C input but not on DC input. Is there are difference?
> 
> I was wondering the same thing, it is so annoying that we can't use a fully charged 6s battery 😑 or maybe we can? Is there anything on the DC path that could not handle 25.2 volts?

I asked the miniware, they say it can handle 28v

**Comment by River-b** [comment]
> I asked the miniware, they say it can handle 28v

they said for  the DC barrel you can use  up to 28V  (like from a Bench DC power supply)?   some DC barrel bricks can spike high and then it's not a good idea  to  use more  than 24V  on DC barrel side. 
or do they mean for USB-C only on EPR?

**Comment by River-b** [comment]
> Let me know if you need any of ny help.

see this  https://github.com/Ralim/IronOS/issues/1420#issuecomment-1334529802

**Comment by PauliusMorku** [comment]
Do all of TS101 come with STM32 chip? Or some of them use clones? I had TS100 with a clone ship and that prevented me to upload IronOS.

**Comment by MaximusPanin** [comment]
> Do all of TS101 come with STM32 chip? Or some of them use clones? I had TS100 with a clone ship and that prevented me to upload IronOS.

Chip clones on the TS100 have appeared recently. This is a strange decision if it is dictated by economy. If the situation with TS101 is similar, they will lose part of the audience. Reputation is hard to build and easy to lose.

**Comment by Ralim** [comment]
I can almost guarantee that they will use clone chips at some point, with all its issues, since its designed using the same STM chip.
The only question is _when_ not if.

**Comment by L1cardo** [comment]
> I can almost guarantee that they will use clone chips at some point, with all its issues, since its designed using the same STM chip. The only question is _when_ not if.

Is there a way to know if it is a clone chip without open the iron case?

**Comment by Ralim** [comment]
Only other hint is usually "is their DFU more dodgy than usual"; generally opening case is easiest to go by until we know other means. (For TS100 we know all made after earlier this year are clones, and DFU versions above a certain number are all clones).

**Comment by MaximusPanin** [comment]
GD32F107 https://youtu.be/Lo77r6JwimU?t=1149
Compare board https://youtu.be/Lo77r6JwimU?t=850

GD32F103TBU6 _(TS101)_ -  **108** MHz
STM32F103T8U6 _(TS100)_ - **72** MHz

**Comment by PUF52** [comment]
Greetings to all lovers of soldering irons ts. The question is addressed to those who have ts101.
The miniware company promised everyone that you can easily replace the regular logo with your own. How to do this is written in the user manual. I tried, but it didn't work out right. My logo appeared, but the miniware logo did not disappear either. That is, at first, when loading, I see the miniware logo, and then my own. Correspondence with the seller did not help. Here is the last answer:
The factory said this is a normal phenomenon, there is no way to solve it for the time being. Will not affect the normal use of the soldering iron.
The conclusion follows from the answer, the developers did not implement this feature in the firmware, but promised ..
Which of you managed to make the correct logo replacement and how?

**Comment by MaximusPanin** [comment]
**PUF52**, firmware 2.0 ?

**Comment by PUF52** [comment]
Yes, now 2.0.

**Comment by MaximusPanin** [comment]
Send me your logo, I'll try after the new year. 
Happy New Year!

**Comment by discip** [comment]
As in the previous miniware irons, part of the logo is in the bootloader.
So the bootloader has to be changed or rewritten.
@Ralim already did that [IronOS-dfu](https://github.com/Ralim/IronOS-dfu).
But I doubt this is going to work with devices with GD chips.

**Comment by PUF52** [comment]
![image](https://user-images.githubusercontent.com/86562320/210164413-81929997-0bba-4aa9-be2d-5e6982849872.png)

**Comment by PUF52** [comment]
> Пришлите мне свой логотип, попробую после нового года. С Новым Годом!

А куда прислать? Дайте свою почту, вот моя: pershin 52@list.ru

**Comment by L1cardo** [comment]
Sorry to bother! Any progress in this?

**Comment by Ralim** [comment]
TL;DR: No, and I have no timeline.


Miniware have gotten less and less helpful as time has gone on.
No schematics or pinmaps or any help anymore really.

I did an initial investigation to figure out where I2C was, and it's been placed on pins that can't do hardware I2C.

This is super demoralising, as it means that a fair bit more work will need to be done for this to implement it all in software and make sure things work.

Can it be done: totally, but it will take a lot more work. And their dumb as a rock bootloader makes it even worse.


I did scratch out some basic support using the replacement bootloader but it would require flashing using an SWD adaptor to use, so it's not a good way forward.

Next steps is that all the signals on the board would need to be traced, and then all the components measured to figure out scaling values.

Miniware are not making this easier, I have asked for schematics and been told NO, I've asked for a basic pinmap and been told NO, I've for any help they can given and been told "nothing is available".
So only real option left would be a lot of hours either decompiling their firmware to tracing with a multimeter to pin it all out.

Which is a ton of work, and not very motivating or interesting. Couple that with (relatively) low demand and impact compared to other things I could spend my time on.



So this is in the bucket for "future" when other projects are done.


I would happily take a PR, or even just a schematic and then I would look at it again. Or even just a roughly documented pinout.


If someone in the community wants to reverse engineer their firmware or the PCB back to a pinmap I would be happy to pick this up again sooner. But until then, burning 100's of hours on this feels like a poor way to spend my time when I could achieve a lot more in other places for the same time.

**Comment by L1cardo** [comment]
I asked a staff of MiniWare, and he said they offered some to you(don't know if it is true). I also asked him if he can offer more, he said he will look into it.
Following are the screenshots of the chat.

![image](https://user-images.githubusercontent.com/33802186/214756948-32b043b1-6cc7-45c7-a0e7-f47579698d6a.png)
![image](https://user-images.githubusercontent.com/33802186/214757179-66566890-3cac-4d83-a4a7-4ed5f974e0e7.png)
![image](https://user-images.githubusercontent.com/33802186/214757354-d5c66d46-6571-4867-9f96-e91d80669c45.png)

**Comment by Ralim** [comment]
They sent me a single TS101 unit (with NOT FOR SALE lasered on it 🤣 ) but refused any supporting documentation when I asked.

**Comment by Wuthoqquan** [comment]
Hi all! Hope this is not going to be too much off topic!
I'd like to have a better understand on how to use the 90W support made available to the TS101 with the latest FW2.0 update.

The question is if the 90W can be unlocked on the TS101 only with a charger that complies with PD3.1 standard and can output 28V, or if it possible to supply the required power also with a PD charger that has for example an output of 20V at 5A (100W).

It is really not clear to my if the TS101 with FW2.0 can reach the advertised 90W power only with PD3.1 chargers (for example: 28V at 5A = 140W), or if it is also perfectly fine to use a PD charger with a maximum rating of 20V at 5A = 100W.

It is really not clear to me how the power management has been implemented in FW2.0 and if the 90W power can only be achieved with the latest PD3.1 chargers.

Thanks a lot!
Marco

**Comment by River-b** [comment]
> The question is if the 90W can be unlocked on the TS101 only with a charger that complies with PD3.1 standard and can output 28V, or if it possible to supply the required power also with a PD charger that has for example an output of 20V at 5A (100W).

You will have to ask  **Miniware** how to do this  as you are asking about  Miniware's own company firmware  and not  Ralim's IronOS firmware.   There is no Ralim's IronOS firmware for the TS101 and this is not miniware support page.   This is open source Ralim's IronOS  which  is able to be written  to  some Miniware  devices   but the  new ts101  only  has  miniware's  official firmware which  you have to talk to Miniware for support.

This ticket should probably be closed  as  Ralim  indicated above  he is probably not going to put more effort into  trying to get  IronOS to work on  TS101 because Miniware has made it too hard to do so.

Ralim said above  in 2 places, he is probably not going to work on firmware for TS101  because Miniware is not cooperative enough and made a mess of things such that reverse engineering it is going to take too much time.
see here  https://github.com/Ralim/IronOS/issues/1420#issuecomment-1334529802
and  here https://github.com/Ralim/IronOS/issues/1420#issuecomment-1404510840

**Comment by River-b** [comment]
> It is really not clear to me how the power management has been implemented in FW2.0 and if the 90W power can only be achieved with the latest PD3.1 chargers.

it's not clear to anyone how Miniware implemented it since the source code for the TS101 is not open as far as I know. See if  Miniware will release the code for the TS101 firmware and email their support to find out.  
Since it's not open, no way to know and only they can answer this.

**Comment by River-b** [comment]
if you really want to use a  28V 140W PD3.1 charger,  I recommend  getting an Iron that supports it  using an open  firmware like Ralim's IronOS.  
Pinecil V2 version (with green thub grip)  is the only Iron I know of that both  supports PD3.1 and has been  tested by many people with PD3.1 chargers at this point  and also **can use open source firmware** - Ralim's IronOS firmware as opposed to  a closed proprietary firmware. 
Being able to use open  source firmware  is an important  feature to look for in Iron's if you want to use  Github IronOS on them, not all irons can do this or the manufacturer makes it too hard to bother trying. 
Pinecil V2  is confirmed by many to work with 28V although theoretically the maximum you could get after subtracting out the resistance of the Tip  is  about 124 Watts on Pinecil V2.  and  in practice using official 240W  PD3.1 cable with a PD3.1 140w 28V charger,  I've  tested up to 110Watts  putting Pinecil V2 into extreme  testing.    
Besides the tip  having  resistance which you have to  use  ohm calculator to account for,  also consider there are other things causing resistance too and subtract for that (pcb components, cable itself,  more resistance as pcb gets hotter).  For Miniware,  their tips are about  8.0 ohms but you should check that out yourself using  multimeter, then add  a little more resistance in the calculation for other  little things, like the usb-c external tester and such. 
https://ohmslawcalculator.com/ohms-law-calculator

For Pinecil V2, we subtract  6.2ohms  for ohms calculator because the new V2  has  specially designed Pine64 shorter  tips which are more ergonomic but also have lower resistance.   However the V2 can also use the normal longer ts100 tips as well, in which you use the 8.0 ohms  in the  Ohm calculator.

 whereas  the Miniware  can not use  the newly designed Pine64 short tips and therefore,  just  use  8.0 normal length tips in the calculations.  
 For miniware then the theoretical maximum is  98W but you might not see more than 85W with resistance of other things added in,  but you have to test this for yourself. 
 
![image](https://user-images.githubusercontent.com/97197236/217344244-ef8f5f51-07a5-47b9-82cf-e12bc0535b86.png)

The Pinecil V2 on the other hand  was designed with  lower resistance shorter  tips of 6.2-6.5 ohms in mind (although it can also use  the  older tips of 8.0 ohms as well if you unplug between tip swapping). 
and as you can see in  ohm calculator,  Pinecil V2  much better utilizes  the PD3.1 charger as it has a theoretical  max  of   126W  using  a real EPR PD 3.1  charger  of 28V  and also  using a real  PD 3.1 240W  USB-C cable.   The cable is important,  if people use  old  normal  usb-C cable  they will not get more than 20V  plugged into PD3.1 charger.  that cable has a special emark chip in it which opens  negotiation for 28V to an IRon. 
![image](https://user-images.githubusercontent.com/97197236/217345132-1915916c-08d2-487a-aee1-0a356cdaa937.png)

**Comment by Wuthoqquan** [comment]
@River-b Thank you so much! I was not expecting such a comprehensive reply to my question about the difference between 100W supplied by a charger capable of delivering 28V and 100W supplied by a charger capable of delivering only 20V but at a higher current.
I do not own a certified PD3.1 charger, but a 100W USB-C charger working at 20V - 5A, hence why I was curious to understand if I could possibly still take advantage of the 90W offered by the TS101 with FW2.0.
I do apologize if I asked the question here, as recognized in my earlier post, where I admitted my query could be off topic.
But I was asking just from a general technical point of view, since I really do not know if 100W can be different, depending on the different combination of originating voltages and currents.
In my specific case, I’ve purchased a 100W PD charger that can supply a maximum of 20V at 5A and I’m now questioning whether I would be better served by a certified PD3.1 charger capable of delivering up to 28V.
I appreciated your insights about the need of proper cables to anyway be able to fully utilize the higher 28V attainable by PD 3.1. Thanks for that as well.

**Comment by River-b** [comment]
> but a 100W USB-C charger working at 20V - 5A,
@Wuthoqquan 
unfortunately   20V 5amp  100W charger will not make difference.  if you look at the Ohm Calculator math you will see this.
Enter simply  20V  and  8ohm  for the Tip.  and you  get  your answer on maximum watts.
then "try"  to clear  ohm calculator  ,  enter  20V,   5Amps,  8 ohm resistance.   you will see  the science auto corrects and you have exact  same  Power watts.

In this situation, you need to  increase the Voltage as  increase of  amps  rating of your charger doesn't really  do anything significant  as you can see for yourself if you play with  Ohm calculator online.

you might try  20V 5amp  100w charger with a 100W  PD usb-c cable and  get a tiny bit more  than using USB-C   normal  PD65W cable,   but I'm  not sure  it will even show on  the miniware device.
I have tested this on Pinecil devices myself  and  no matter if I use    different cable  PD100w cable or PD60 cable with  either a PD65W 20V 3.25amp charger or a PD100W 20V 5amp charger,   the heating time and performance and  also watts  used  is approxiamately the same in real life.

The only true significant change I could see and feel in performance was upgrading to full 28V  charger  PD3.1 and combining this  with  a PD240W  EPR PD3.1 cable  ( many  work  you dont need  certified ones).  I recently got some  $5  cables from aliexpress of  PD3.1  240w kind and they work  fine  just like my  $15 -$50  USB-C  PD3.1 cables.   for the charger,  if you want a budget one that has proven to be  pretty good and passed  test,   the Rocoren  140w 28V one  has been tested by many and is about  $44  on Aliexpress (use store coupons)  and also  is on  Amazon too.

**Comment by Wuthoqquan** [comment]
@River-b WOW…. Thank you! Hats off! Your reply was again super informative! Now all makes sense to me. The way you have explained how the voltage impacts on the overall supplied power has clarified why PD 3.1 should be the preferred option in such use cases, given all the other variables that reduce the actual power output.
Guess I will send back the 100W charger and replace it with a proper PD 3.1 140W one.
Many thanks again!

**Comment by River-b** [comment]
@Wuthoqquan 
PD3.1 is also more future proof.  and the USB-C  240W cables also future proof.
All  100W cables were officially deprecated by the USB-IF  organization in December 2021  because they want to simplify cables into just 2 kinds. One category is  SPR ( standard power)  PD60 cables - cheaper because  no emark smart chip inside.   and the 2nd category is  EPR  PD3.1 cable (extended power  rating)  which is  the 240W cables that come in different data speed.  for  Firmware  even the lowest 480 mb/s  is fast enough data speed so we only care about the 240W rating of the cable.
Be aware there are some fake  240W cables on the market  that I purchased and others as well, and tested.
I have both certified and non-certified.
following cheap cables worked at  28V  on usb-tester for me if you want to save some money.

- https://www.aliexpress.us/item/3256804889230539.html
- https://www.aliexpress.us/item/3256804901287548.html

they are not certified by USB-IF but they work and show me correct 28V on USB-C tester and on Pinecil Iron Detailed screen view. 

yes 100W cables are still sold. I dont see reason to buy them unless you get it dirt cheap same price as SPR PD60 cable.  USB-IF said manufacturers who have certification on 100W cables can continue making them, but  when they run out of  100W emark chips, they too will probably switch to selling  EPR 240W emark cables.
reference:  https://compliance.usb.org/index.asp?UpdateFile=Cables%20and%20Connectors&Format=Standard#:~:text=Effective%20Immediately%20100%20Watt%20USB,on%20December%2022nd%2C%202021.

**Comment by rozwell** [comment]
> This ticket @rozwell - maybe consider closing as Ralim indicated above he is probably not going to put more effort into trying to get IronOS to work on TS101 because Miniware has made it too hard to do so.

Looks like TS101 is a no-go :unamused: 
And this thread went off-topic so... I agree, closing.

**Comment by violavon** [comment]
@rozwell please, re-open.

**Comment by piit79** [comment]
@violavon Why...? It was explained adding support would be complicated, and there is no plan to do it. You can blame Miniware for that for being ignorant.

**Comment by violavon** [comment]
@piit79 As said, support may be done in the future, although it isn't a high priority.

**Comment by River-b** [comment]
and it will be basic because... read all comments from ralim in this ticket, miniware has laid it out in a way that it's too much work to reverse engineer and they haven't shared with Ralim the schematics or anything to make it easier.

**Comment by violavon** [comment]
Yep, and until then this should be open.

**Comment by rozwell** [comment]
Oh well, looks like there is still a chance, so let's keep it open until @Ralim decides otherwise ;)

**Comment by L1cardo** [comment]
> Oh well, looks like there is still a chance, so let's keep it open until @Ralim decides otherwise ;)

I don't think so. Those are just some words without any meaning, just to make you go away.

**Comment by bulletng** [comment]
I think I'm going to return my TS101 and upgrade to the Pine64 because of this!

**Comment by gruvw** [comment]
Got this response:

![IMG_20230427_073516.jpg](https://user-images.githubusercontent.com/63407038/234769065-61dacf51-3ea6-4a8a-bacf-6eff192407bb.jpg)

Don't really know what it means 🤷‍♂️

**Comment by Ralim** [comment]
Neither do I so far 🤷🏼

**Comment by whitequark** [comment]
Hey folks, I've talked to Ralim and I am willing to do the hardware analysis work necessary to get TS101 supported. I think I'll have to sacrifice one unit for it, are any of you willing to buy me one after I post the reverse-engineered documentation necessary for bringup?

**Comment by L1cardo** [comment]
How about we all pay some?

**Comment by whitequark** [comment]
As long as someone who isn't me coordinates it :)

**Comment by L1cardo** [comment]
Is there any website that can do something like raising money?

**Comment by whitequark** [comment]
@Tasshack I talked about what would motivate them to add support for TS101, and they replied that getting a pin map (mainly heater and the USB PD PHY I2C) would go a long way. It's not really difficult for me to take my TS101 apart and reverse-engineer it to a sufficient degree, so I'll just do that, I'm thinking tomorrow or the day after.

**Comment by whitequark** [comment]
# How to tear down your TS101 (and then be able to put it back in one piece)

This is actually not very difficult, and consists of only a few steps. To do this you will need three tools: a screwdriver set, a box cutter or any other narrow bladed knife, and a plastic opener tool of the kind they sell with smartphone displays (or ideally two).

**Read all of these steps before starting to do anything.** This minimizes the chance of breaking your iron by accident.

![PXL_20230602_211840265](https://github.com/Ralim/IronOS/assets/54771/36b9dee2-dcef-4b34-9fa3-377f4b149f69)

## 1. Removing the glass

First of all you will need to remove the glass in front of the display. As best as I can tell, **This is an actual glass!** It does not seem to be an acrylic or polycarbonate panel. **Do not put excess strain on it, or you will crack it.**

The glass is affixed to the case with two small cuts of adhesive strip under the darkened areas, one in front and one in back of the soldering iron. Use the knife to pry the glass on the **short side of the glass, close to the DC barrel jack**. This will damage the plastic a little, but nothing too noticeable.

Once you have your knife wedged in between the glass and the case, you have separated the glass from the first adhesive pad (or the pad from the case). **Now, use any tool** (a fingernail works) **to lift the glass from that point**. Try not to put fingerprints of it, or you'll hate yourself after reassembly. Once it is lifted enough, **grab it by the sides and remove it**. It is strong enough to withstand a fair bit of force as long as it is not bent.

## 2. Disassembling the case

The case is held together with four latches, indicated below. There are no fasteners; it is just plastic.

![PXL_20230602_211919772_edit](https://github.com/Ralim/IronOS/assets/54771/3434ccdf-1fdd-4496-91a6-4ab0f540d809)
![PXL_20230602_211907241_edit](https://github.com/Ralim/IronOS/assets/54771/16983840-f734-4633-b2ff-1a4361858888)

To start the disassembly, use a thin soft object (a plastic opener or a fingernail all work) to **lift the bottom part of the case at the *short* segment of the seam closest to the DC barrel jack**. The plastic is thin and flexible and the nearest latch is far away, so it is not hard to separate it maybe 0.3-0.5 mm from the other part of the case. This causes the *long* segment of the seam, parallel to the case, to open up a bit. While holding the case in this position (it helps to squish it from the sides with your fingers), jam the plastic opener tool **into the seam close to the DC barrel jack**. Now that it is inside, drag it across the length of the seam, undoing the latches. Repeat on the other side.

## 3. Remove fasteners

**Remove the hex head grounding screw on the outside of the iron, and the Philips head screw inside of the iron where the contacts are.** You can leave the slot head screw in place, it doesn't matter. Also, **remove the small half-circular PCB** that is installed upright inside the iron, it is not held by anything, just inserted into the slot.

## 4. Starting extraction of the PCB

Take the thickest hex or Torx bit from your screwdriver set and put it into the screwdriver. Alternatively, something like a retractable pen would work (and may be even more convenient). Grab the case in one hand and your screwdriver or pen into another. **While holding the case firmly, push the central pin of the DC barrel jack with the object you're holding, until the PCB moves a little down the case.** The whole assembly will flex: the case and the PCB both. This is normal. Try not to break anything. When the assembly is flexing enough to have clearance for removing the DC barrel jack from the case, push it up and out of the case.

Once you're done, the iron should look like this:

![PXL_20230602_212114173](https://github.com/Ralim/IronOS/assets/54771/8c6289f6-7219-4ad6-a42b-7be1fb4efefc)

**Note the display held by my finger**. This is important for the next step.

## 5. Finishing extraction of the PCB

As can be seen on the final image below, the case is sandwiched between the display and the PCB. There is an opening through which the flex cable passes, and the display can only be removed if it is tilted at an angle to the case. This is why the first step is removing the glass.

**While holding the display at about 45° angle, start carefully pulling the PCB out of the case.** The goal is NOT to completely remove it yet, but to move it to the point where there is more resistance to pulling it. Once this happens (and the PCB is stuck at the part of the case where latches are molded into it), you can release the display, since it will hang at an angle by itself. At this point, **use both hands to pull the case and the PCB apart**. This takes a little bit of force, but not too much, and the flat cable of the display should be reasonably safe from tearing at this point.

**You're done!** Put the screws and the display into a plastic bag and get yourself a coffee--the process was probably tense enough to warrant it.

![PXL_20230602_212237971](https://github.com/Ralim/IronOS/assets/54771/8d99ca54-26d7-4cae-ac0d-af53238bf6ed)

**Comment by whitequark** [comment]
# High resolution images of the PCB

Top 1

![PXL_20230602_215953670_crop](https://github.com/Ralim/IronOS/assets/54771/5f92a000-7d07-44f4-9a5c-2d46facb8b5c)

Top 2

![PXL_20230602_220120595_crop](https://github.com/Ralim/IronOS/assets/54771/6a3843b5-96b2-49de-85f1-5289958a5ebb)

Bottom

![PXL_20230602_220149601_crop](https://github.com/Ralim/IronOS/assets/54771/4c51b451-4958-449e-bb14-9aa22e98dcb6)

**Comment by whitequark** [comment]
# Composite image of both PCB sides, to scale

![TS101_PCB_composite](https://github.com/Ralim/IronOS/assets/54771/60f52625-5280-4ab8-be7d-f80244677342)

**Comment by whitequark** [comment]
# Pinout (in progress)

## MCU pinout

**NRST**:
**BOOT0**: Vss
**PA0**: 
**PA1**:
**PA2:**
**PA3:**
**PA4:**
**PA5:**
**PA6:**
**PA7:** display pin 9
**PA8:**
**PA9:**
**PA10:**
**PA11:**
**PA12:**
**PA13:**
**PA14:** n/c
**PA15:**
**PB0:** display pin 10
**PB1:** display pin 11
**PB2:** pull-down to Vss
**PB3:**
**PB4:**
**PB5:** USB-PD SDA
**PB6:** USB-PD SCL
**PB7:** n/c
**PD0**: n/c
**PD1**: n/c

## Display pinout

The pin 1 of the display is towards the same side of the PCB as the pin 1 mark of the MCU (on the bottom, if the text on the MCU is oriented normally).

The display is almost certainly connected via I2C and is likely [SSD1306](http://www.adafruit.com/datasheets/SSD1306.pdf) compatible.

**14**: capacitor to Vdd
**13**: capacitor to Vdd
**12**: pull to Vss ?IREF?
**11**: PB1
**10**: PB0
**9**: PA7
**8**: Vdd
**7**: Vss
**6**: n/c
**5**: Vdd
**4**: C2x
**3**: C2y
**2**: C1x
**1**: C1y

x,y = P,N or N,P

**Comment by whitequark** [comment]
The pinout above (the third oldest edit) is probably about as far as offline tracing can go; further investigation requires firmware reverse-engineering and/or online tracing (using a debugger to toggle pins and seeing what happens).

**Comment by whitequark** [comment]
A brave type-C USB cable has been chosen to voluntarily give its life in the service of the Greater Good. Please take a moment of silence to honor its brave sacrifice, which will forever live in our hearts. 🫡🎖️🫡

I can confirm that the programming fixture that was speculated above works.

![PXL_20230603_005442897](https://github.com/Ralim/IronOS/assets/54771/cf683620-e66d-4752-90fb-0efb10bf516c)

![image_2023-06-03_01-50-39](https://github.com/Ralim/IronOS/assets/54771/2914ce63-87c6-41d6-8507-717f381b100a)

**Comment by whitequark** [comment]
# Stock firmware image

[Download (128 KB binary dump)](https://github.com/Ralim/IronOS/files/11641847/ts101_factory.zip)

**Comment by whitequark** [comment]
@Ralim Aside from **PA1** and **PA5**, which neither have a confirmation for their function nor a definitive exclusion of any, the above is a complete pinout that has been verified against both the stock firmware and the hardware. I hope this helps!

By the way, most of the credit goes to @VioletEternity—I just gave her the iron, asked for help, and communicated the results. The whole process took her about 8 hours, which seems a little bit less than _hundreds_ :p

**Comment by Ralim** [comment]
@whitequark & @VioletEternity 
Thank you so much 🙇🏼 

This not only gives me all the missing pinout information, you even traced out the extra functions for each pins which confirms their firmware is running a lot like it used to (before it went effective closed source).

Nice to see no real innovation going on here, so everything apart from higher res screen should be quite quickly to bring up. But also _whyyyy_ do they hate hardware I2C so much. Same issue as S60 where we spend a bunch of time bit-banging I2C rather than being productive.

I'll try and get on porting this over asap.

**Comment by VioletEternity** [comment]
Some additional information that may be useful.

## Display controller ASIC identification

I have not verified this myself, but in [a comment](https://github.com/Ralim/IronOS/issues/1420#issuecomment-1367014226), a review was linked which describes ([timecode](https://www.youtube.com/watch?v=Lo77r6JwimU&t=1149s)) the display as being based on  the [SSD1316](https://www.orientdisplay.com/wp-content/uploads/2022/08/SSD1316.pdf) controller.

## Accelerometer ASIC identification and interface

I reverse-engineered the I2C initialization code for the accelerometer. Based on that and the package size, the cheapest (on Mouser) accelerometer that has a register set that is compatible with the initialization code would be [ST LIS2HH12](https://www.mouser.co.uk/datasheet/2/389/lis2hh12-1849522.pdf). The actual device is different; both the I2C address and the value of the WHO_AM_I register are different. There are many accelerometers with this type of interface and it is probably not important exactly which device it is considering it is only used to detect lifting the iron.

The annotation accelerometer initialization and readout code follows:
![image](https://github.com/Ralim/IronOS/assets/96890070/ea9fb7f1-2826-4e7f-ba79-244b480c5c83)
![image](https://github.com/Ralim/IronOS/assets/96890070/929c4122-09ab-40cb-8e6c-0041d5e00bd9)
![image](https://github.com/Ralim/IronOS/assets/96890070/8ab516f3-d2d0-4f7a-b131-1cfd11702e2d)

This code is extracted from the bootloader's accelerometer test routine.

This function confirms the pinout:

![image](https://github.com/Ralim/IronOS/assets/96890070/8d2cb563-127f-4599-bb39-bb925219e6c6)

## Tip temperature measurement

The temperature measurement is performed twice, with the **PA1** pin (cc @whitequark to update the function table) used to, most likely, multiplex something. It is being toggled in `GPIOA->ODR` using a bit banding area at `0x42210184`:

![image](https://github.com/Ralim/IronOS/assets/96890070/a704f6f3-65f2-443c-88b9-c3b53f97043c)

The `tip_degreesC` variable is computed using software floating point that I didn't completely label. The computation looks like this in pseudocode, with `OPx` representing unknown arithmetic functions:

    void OP1(float,float);
    void OP2(float,float);
    void OP3(float,int);
    void OP4(float,int);
    PA1 = 1
    sample1 = measure_channel(3)
    PA1 = 0
    sample2 = measure_channel(3)
    tip_degreesC = (int)((((((((((float )sample1) OP1 (float)sample2) OP2 3300.0f) OP2 1000.0f) OP2 100.0f) OP3 -12) OP4 2843) OP4 13)

I can dig into this further if necessary, but it seems like @Ralim might know the shape of the expression already.

**Comment by VioletEternity** [comment]
## Addendum: Tip temperature sensing circuit overview

The tip temperature sensing circuit is based on the [RS8551](https://datasheet.lcsc.com/szlcsc/Jiangsu-RUNIC-Tech-RS8551XK_C255431.pdf) operational amplifier and is located in this area of the PCBA:

![image](https://github.com/Ralim/IronOS/assets/96890070/8102776e-04e1-4293-8bd0-ce30d837bca6)

The OUT pin of RS8551 is connected to **PA3** through an 1 kOhm resistor. **PA1** is connected to the anode of D1 (the via is located under D1). The net marked with red is connected to one of the tip terminals (the one closer to the MCU). The other terminal is grounded.

I'm not familiar with soldering iron control algorithms but perhaps **PA1** is used to inject current to measure tip resistance?

**Comment by sandmanRO** [comment]
Hello @VioletEternity, it could be that chip (PA5 connected) is a 3.3V voltage regulator? Otherwise, I could see why (I'm speculating now, of course) someone might want to "measure" the Vdda level. Theoretically this measurement would return the maximum count (or the ADC resolution). This could be useful when computing the actual voltage level on other channels when you don't know precisely if the ADCs you are going to use are for example 12bit or 16bit.

**Comment by sandmanRO** [comment]
> I'm not familiar with soldering iron control algorithms but perhaps **PA1** is used to inject current to measure tip resistance?

If the TS101 is using the same tips as TS100, then the temperature measurement relies on measuring the thermoelectric voltage. I'm speculating again...that current injection is perhaps this part of open thermocouple detection algorithm rather than an effort to measure the temperature based on temperature dependent resistance measurements. But I might be wrong.

**Comment by VioletEternity** [comment]
> rather than an effort to measure the temperature based on temperature dependent resistance measurements

My suspicion is that the tip resistance is measured so that the power can be displayed or limited. There is no circuitry to measure the tip heating current, and given the known voltage, calculating the tip power requires knowing its resistance.

**Comment by sandmanRO** [comment]
Good point. It could be precisely as you said...

**Comment by VioletEternity** [comment]
> Theoretically this measurement would return the maximum count (or the ADC resolution).

The Vdda level is the same as the Vdd level (I don't entirely recall but I think they need to be connected to the same rail with this SoC), and so it should always measure as all-ones. There is no need to expend a pin to find that out.

But if the operational amplifier V+ rail was, in an earlier revision, connected to something other than Vdda, then this pin could have been used as a reference measurement for the operational amplifier output, which is what I was suggesting in my earlier message.

**Comment by VioletEternity** [comment]
## Addendum: Outline of functional blocks

The following is an outline of the PCBA location of every known functional block used in the iron:

![ts101_func_blocks](https://github.com/Ralim/IronOS/assets/96890070/9b2d11a3-d539-4082-af60-c23207fae284)

**Comment by sandmanRO** [comment]
Sure thing, @VioletEternity...what I was trying to say is rather academic: if you are going to use a chipset that embeds a 12bit ADC variant, the maximum count would be 4095. In the same time, a 16bit ADC variant would return a value of 65535. "Measuring" the VDD and using the result for computing the voltage on other channels would allow your code to run without change on both platforms.

**Comment by VioletEternity** [comment]
> if you are going to use a chipset that embeds a 12bit ADC variant, the maximum count would be 4095. In the same time, a 16bit ADC variant would return a value of 65535

Usually this is solved on the STM32 devices by setting ADC_CR2.ALIGN (bit 11) to 1:

![image](https://github.com/Ralim/IronOS/assets/96890070/f0b127f8-cec9-4ddb-8b92-eef002ebd6a9)

I don't believe this is done in the stock firmwar.

**Comment by sandmanRO** [comment]
All right, @VioletEternity. Please disregards and focus on what it matters.

**Comment by whitequark** [comment]
I've updated the pinout to its final version.

Also, lmao @ Miniware not giving @Ralim any information for months because of concerns that a competitor could reverse-engineer it when @VioletEternity reverse-engineered most of the functions (and certainly enough to clone it) in less than 24 hours. All they're doing is draining goodwill.

**Comment by Ralim** [comment]
RE: 
>Addendum: Tip temperature sensing circuit overview

Thank you for this 🙇🏼 
Can confirm that PA1 + Diode + Resistor is how we measure the tip resistance (use this to inject known current across tip heating element, then use the normal op-amp readback to read it. Then toggle a few times to get the delta and convert back to resistance).


>Tip temperature measurement

Thank you for this, I wouldn't dig too much into this algorithum, this looks fairly similar to their existing conversion function that they have used for a while (that tends to fail on clone tips a fair bit). So not worth your time to dig into this  Ithink.



>Accelerometer ASIC identification and interface

Thank you sooo much, hunting for which damn accelerometer they have used _this time_ is always a pain, as they tend to mix-and-match. I'll have to add support for this on to IronOS (yay another one lol)

> OLED

I can confirm its the SSD1306 compatible, I have managed to get a picture up on the OLED by stubbing out most of the rest of the init.



Miniware always answered no when I asked roughly every Month. Even when I pointed out its going to be like 1 weekend for somone to figure out all their secrets. Thank you hugely for all of the information, I have like 70-80% of a port done already.


I'm rather glad you have gone far beyond a schematic and have provided enough details for someone to copy in a way, now its on them for being stubborn 🤣. Ever since I started to support more products than theirs they became insanely stubborn, while all others I've talked to have been at the least _polite_ and generally exceedingly helpful.

Still; huge thankyou.

**Comment by VioletEternity** [comment]
## Replacement display

Although I took care to not damage the display, it seems that my handling caused an invisible crack to appear in the display FPC, and now it works only if the FPC is flexed in certain ways. (This does not include the way in which it's flexed when the iron is assembled.) In retrospect, I should have desoldered it before handling the PCB this much.

In any case, I looked through Aliexpress listings for 0.87 inch OLED displays, and [this one](https://www.aliexpress.com/item/1005005110745146.html) seems to be an exact match. In particular, I confirmed that the dimensions of the glass and the length of the FPC match this drawing (copied from the listing):

![image](https://github.com/Ralim/IronOS/assets/96890070/5f87d724-4a72-476b-bd83-eace6bf828e8)

The display costs 3.24 GBP (incl. shipping), so breaking the display doesn't seem to be an expensive mistake after all.

**Comment by sandmanRO** [comment]
> Can confirm that PA1 + Diode + Resistor is how we measure the tip resistance

Hello @Ralim, thank you turning the lights on in the room. I see now...that "+ Rezistor" makes all the difference...so actually there is a resistor series connected with the diode so the PA1 output goes through a diode AND a known resistor before sinking into the tip and that would allow computing the tip resistance indeed. I noticed in the beautiful addendum @VioletEternity posted that there might be an extra 100k pull-up resistor that would also continuously sink some current into the tip. Is that correct? If so, the resistance calculation would have to be lightly adjusted, I guess. It could also be that the parallel capacitor + 63k resistor would go to amplifier output (as opposed to V+ as stated in the picture) and that would be the negative feedback (the amplification factor would be 63k / 2.4k + 1)?

Hello @VioletEternity, sorry to hear about the display issue. Although the display is not expensive it still sucks when it happens.  
This being said, it would be too much trouble for you the post the full dissembled text? I still have a hard time understanding why they would use both sample1 and sample2 for computing the tip temperature, when normally only sample2 value would be needed. I noticed for example from the short snipped you posted as a picture that while sample1 is plain ADC count, on sample2 they do something with ADC value first, and only afterwards they use the value further on, so sample1 and sample2 would not be quite the same fruit so they could not be simply subtracted in order to compute the tip resistance.

**Comment by VioletEternity** [comment]
@sandmanRO I got it to work again using a bit of glue, but as I'm borrowing the iron from @whitequark I'll fix it properly as soon as I can.

Sure thing. Here's what I have right now:

```c
int32_t tip_measure_temp()

{
    bit_band_ODR_PA1 = 1;
    sub_800a4e8(4);
    uint32_t sample1 = adc_sample_once(3);
    bit_band_ODR_PA1 = 0;
    sub_800a4e8(4);
    float sample2_f = i32_to_f32(adc_sample_once(3));
    int32_t tip_degreesC = ((int32_t)f32_to_i32(sub_80082de(sub_80082de(sub_800835a(sub_800827a(sub_800827a(sub_800827a(sub_800826e(i32_to_f32(sample1), sample2_f), 3300), 1000), 100), -12), 2843), 13)));
    if (tip_degreesC < 0)
    {
        tip_degreesC = 0;
    }
    if ((tip_degreesC >= 0 && tip_degreesC > 400))
    {
        data_2000001e = 1;
    }
    if ((tip_degreesC < 0 || (tip_degreesC >= 0 && tip_degreesC <= 400)))
    {
        data_2000001e = 0;
        data_20000023 = (data_20000023 + 1);
    }
    if ((((uint32_t)data_2000001e) != 1 || (((uint32_t)data_2000001e) == 1 && ((uint32_t)data_20000022) != 1)))
    {
        data_20000021 = 0;
        data_20000030 = 0;
    }
    data_20000022 = data_2000001e;
    uint32_t r1_6 = ((uint32_t)data_20000030);
    *(int16_t*)(0x2000032c + (r1_6 << 1)) = ((int16_t)tip_degreesC);
    uint32_t r0_11 = ((uint32_t)(r1_6 + 1));
    data_20000030 = ((int16_t)r0_11);
    if (r0_11 >= 5)
    {
        data_20000030 = 0;
        data_20000021 = 1;
    }
    int32_t r0_12 = 0;
    uint32_t r1_9 = 0;
    int16_t r0_14;
    if (((uint32_t)data_20000021) == 0)
    {
        uint32_t r2_2;
        for (r2_2 = ((uint32_t)data_20000030); r1_9 < r2_2; r1_9 = ((uint32_t)(r1_9 + 1)))
        {
            r0_12 = ((int32_t)(r0_12 + *(int16_t*)(0x2000032c + (r1_9 << 1))));
        }
        r0_14 = ((int16_t)(r0_12 / r2_2));
    }
    else
    {
        do
        {
            int16_t r0_13 = (r0_12 + *(int16_t*)(0x2000032c + (r1_9 << 1)));
            r1_9 = ((uint32_t)(r1_9 + 1));
            r0_12 = ((int32_t)r0_13);
        } while (r1_9 < 5);
        r0_14 = ((int16_t)(r0_12 / 5));
    }
    data_2000002e = r0_14;
    return 1;
}
```

**Comment by sandmanRO** [comment]
Thank you very much @VioletEternity! It looks like the sample1 value is also converted to float when passing its value as function argument so inside that function both sample1 and sample2 values would both be apples after all.

**Comment by L1cardo** [comment]
# **IMPORTANT！！！**

**DO NOT ERASE DFU!**
**DO NOT ERASE DFU!**
**DO NOT ERASE DFU!**


I have contacted Miniware, asked them since the whole thing is clear now, can they provide an official pinout and schematic?
They are still saying some BS, except this: **DO NOT ERASE DFU!**

It will brick even if we backup the DFU and then flash the DFU later.

![image](https://github.com/Ralim/IronOS/assets/33802186/ea1dd586-26b0-45e9-8431-ec31a7c44c4e)

**Comment by Ralim** [comment]
Their bootloader _doesnt let you erase it_. (Unless it has new, exciting bugs to add to the hundreds of tickets their "amazing" 🙈  bootloader has caused me to have to deal with).

I have seen this being said so many times and I have no idea why really; its quite trivial to re-load their DFU from a backup you have taken to restore. _BUT_ you cant load someone elses backup, as like the MHP30 they are serialised.

But dont worry, I'm working on the IronOS-Dfu supporting this, so you wouldn't ever get a brick. You just worse case miss out on being forced to view the Miniware Logo on boot each time.

**Comment by L1cardo** [comment]
> its quite trivial to re-load their DFU from a backup you have taken to restore. BUT you cant load someone elses backup, as like the MHP30 they are serialised

Yeah, this is what I am thinking, as long as you backup your own bootloader, there won't be issue.

**Comment by VioletEternity** [comment]
> Brother, can you help me to remind them? The DFU of TS101 must not be erased, otherwise TS101 will be GG(game over), even if you read another DFU and then flash into it, it is useless, you can only send it back to the factory to recover.

Or you could (the firmware dump I posted) change the byte at offset 0x542c from 0xb3 to 0xb4 and the DFU will boot on any device. The constant to look for in the disassembly if you want to patch a different DFU version is 0x1ffff7e8 (the UNIQUE_ID register).

**Comment by sandmanRO** [comment]
The only thing that I could think of is sample1 and sample2 are used to calculate the thermocouple's TEM (thermoelectromotive) voltage out of the offset induced by that 100k pull-up resistor. The tip resistance would be calculated along the way. When PA1 is high, the equations would be like these:

VDD = I1 * R100k + Rtip * (I1 +I2) + TEM (1)
VDD = I2 * R24k + Vdiode + Rtip * (I1 +I2) + TEM (2)
sample1 = Rtip * (I1 +I2) + TEM (3)

where I1 is the current that flows via 100k resistor, I2 is the current that flows via 24k resistor, Vdiode is the diode forward voltage, Rtip is the tip resistance and TEM is the actual thermocouple output. We have 4 unknowns, I1, I2, Rtip and TEM. We need one more equation. When PA1 is low we have:

sample2 = VDD * Rtip / (Rtip + R100k) + TEM (4)

From eq. (1), (2) and (3):
I1 = (VDD - sample1) / R100k
I2 = (VDD - Vdiode - sample1) / R24k

Let's have Itip = I1 + I2, by now a known value. Also, Rtip is significantly lower than 100kohm so Rtip + R100k ~= R100k. The eq. (3) and (4) would become:

sample1 = Rtip * Itip + TEM (3')
sample2 = Rtip * VDD / R100k + TEMP (4')

Subtracting (4') from (3') would lead to:

sample1 - sample2 = Rtip * (Itip - VDD / R100k) 

so

Rtip = (sample1 - sample2) / (Itip - VDD / R100k) 

So now Rtip is a known value as well. Then, using (3') we can also compute actual TEM:

TEM = sample1 - Rtip * Itip

**Comment by VioletEternity** [comment]
> Or you could (in the firmware dump I posted) change the byte at offset 0x542c from 0xb3 to 0xb4 and the DFU will boot on any device.

Apologies, that isn't quite right; I missed a few other places. Here's a patched DFU that will work on any device: 
[ts101_dfu_patched.zip](https://github.com/Ralim/IronOS/files/11651179/ts101_dfu_patched.zip)

**Note:** I only patched the bootloader. The stock firmware also performs the same UNIQUE_ID check, and this DFU will *not* let you run Miniware's stock firmware at all.

**Comment by VioletEternity** [comment]
Incidentally, while looking through the copy protection code, I found out what **PA5** and the corresponding 2111/2225 device is for. There is a function in the bootloader that manipulates that pin and if the outcome isn't right, it displays "IDChip Err". Considering that what's connected to **PA5** is a two terminal ASIC, it seems safe to assume this is a 1-Wire device or something similar, though I haven't reverse-engineered the protocol in detail and it doesn't look very much like 1-Wire to me.

**Comment by gamelaster** [comment]
@VioletEternity so this is counterfeit measurement? :D

**Comment by sandmanRO** [comment]
The more I think this through the more I am convinced that the design has nothing to do with tip resistance measurements but rather an open thermocouple detection (ODT) that went overkill, that is they simultaneously used a static / passive OTD (via that 100k resistor) and also that dynamic OTD (PA1 + Diode + Resistor). The issue is, once you use a dynamic OTD there is no point in going redundant with a passive OTD and vice-versa. Moreover, the passive OTD shall NOT inject more than few nA into the thermocouple (TC) so the induced bias would significantly influence TC's TEM. Well, that is not the case here. That 100k pull-up would alter the results on a TS100 tip in excess of 12-13°C. 

Let's try to broaden the picture here: do we really need to measure the tip resistance? I think not. The Miniware T100 tips seem to be fairly consistent in terms resistance. I have a whole bunch of TS100 Miniware tips, from pin-point tips to the large horse-shoe. Their resistance slightly varies within 7.5-7.7ohms. Beyond that, the IronOS does not report the injected power in absolute terms but as a relative value, and for that you don't really need to know the tip resistance as the value comes out from the "duty-cycle" value of power modulation pulse. Although I do not have a TS101 (nor intend to have one for that matter as I'm happy with a TS100s I have already), I am told that TS101 does not report the output power in absolute terms either, but still as a relative power bar, same as IronOS does.

Ok, let's assume we really WANT (as opposed to NEED) to measure the tip resistance using the math described above. We need to keep in mind that IronOS is using a hardware paced (hardware timers driven) ADC as opposed to the lame software polled ADCs used with Miniware firmware. To get this working, a new ADC injected channel would have to be configured in such a way it would be triggered when the PA1 rising edge while PA1 would have to go high precisely at the end of the power pulse gate, then go low, when a second ADC injected channel would be triggered, this time by the falling edge PA1 output, while PA1 output would be driven by yet another timer daisy-chained as slave to the power gate timer. Really? What a mess...

Stay tuned as that's not all. Let's remember that all the data acquisition stuff has to fit in a very narrow window of time, between two consecutive power modulation pulses. The operational amplifier that handles the tip's thermocouple output has a capacitor on the negative feedback...good for filtering out noises but that also induces a small delay, so every time PA1 changes state, we need to wait a bit before we could take a temperature measurement. I don't know what the negative feedback capacitor value to calculate the precise delay needed, but if delay is in range of ms, the entire IronOS' ADC timing would have to be reconsidered, basically by slowing everything down and that would not be a good thing as it would affect the PID performance.

Perhaps the best approach to measuring the resistance is to not do it continuously but only once per heating session, right before the session starts.

Of course, @Ralim would decide if he wants to go ahead with this, but my suggestion would be to let IronOS as it is. That 250µV offset induced by the 100k pull-up resistor (+10°C offset) could be easily corrected via the temperature calibration offset, and the mechanism for that is already in place. In the end, let's remember that we are talking about a damn soldering iron and not a precision temperature measurement device.

**Comment by Ralim** [comment]
>Perhaps the best approach to measuring the resistance is to not do it continuously but only once per heating session, right before the session starts.

This is what I'm currently doing for PinecilV2, where I measure once at boot and roll with that. Since also keeping in mind if you change the resistance you (might) need to re-negotiate USB-PD. At that point its just easier to power cycle.
For TS101 I'll probably just port the same tip measurement code over so it can be used with the Pine64 short tips (lower resistance).

>Of course, @Ralim would decide if he wants to go ahead with this, but my suggestion would be to let IronOS as it is. That 250µV offset induced by the 100k pull-up resistor (+10°C offset) could be easily corrected via the temperature calibration offset, and the mechanism for that is already in place. In the end, let's remember that we are talking about a damn soldering iron and not a precision temperature measurement device.

Since nearly all the units out there have this ~100K ish pullup resistor its similar for most. The current code assumes a certain uV offset on tips, included in that is this resistors impact. But yeah we are making a soldering Iron not some $200 measurement equipment 🙃

**Comment by sorgelig** [comment]
> Since also keeping in mind if you change the resistance you (might) need to re-negotiate USB-PD.

what's problem to request maximum available current but use less current? In this case iron won't need to re-negotiate PD and simply use what is available.

**Comment by Ralim** [comment]
These tips are only resistive, and to meet PD specs; we can't draw more than the negotiated current.

So the problem happens when you have a low resistance tip, it can mean that you can't negotiate the max voltage. For example 60W capped chargers for 20V,3A. If you have a 6 ohm tip, that will exceed the 3A and so you need to drop down to 15V.

We can't pwm the tip anywhere near fast enough to just the pwm to regulate the current, nor do we have an inductor to filter it.

**Comment by Hoholok21** [comment]
@Ralim Hello there! Don't think it's arrogance, but could you tell me how many percent of the firmware is ready?

**Comment by iLLiac4** [comment]
43,37%

**Comment by Ralim** [comment]
@Hoholok21 
Generally asking a question like this is ignored, especially when the linked PR has a progress / todo list on it that tells you what does and does not work. Additionally one could read the commit stream there to see whats being worked on at that point in time.


However, I _would_ like to update that #1695 _should_ be around about a testable working image now.
I _think_ I have the core parts done (sans the OLED at full resolution), which should allow for hardware testing as I'm sure there will be bugs or missing features.
Bugs and missing features should be noted on that PR (not here).

**Comment by VioletEternity** [comment]
For attaching the display cover glass I have found that [this product](https://www.amazon.co.uk/gp/product/B07K6PJ9SK/) works well. It is very thin, appears to be strong enough for the task, and is fairly cheap.

**Comment by Ralim** [comment]
Going to close this as done. All the hardware support for TS101 is now in `dev` and future releases.
Full larger OLED support is coming but will be a future release.

**Comment by whitequark** [comment]
@Ralim Thanks so much! I'm happily using my TS101 now \o/

**Comment by Ralim** [comment]
Please as always feel free to open issues should things break of course. Otherwise, happy soldering 😜

**Comment by RadioTek67** [comment]
Hi, @Ralim! Where can I get a your file for TS101 ?

**Comment by gamelaster** [comment]
@RadioTek67 download latest dev CI build here: https://github.com/Ralim/IronOS/actions/runs/5643689459

**Comment by rozwell** [comment]
It was brought up before that input voltage can be 28V for DC (what MINIWARE officially confirms).
I stumbled upon their comment [here](https://oscarliang.com/ts101-soldering-iron/#comment-158612) where they claim upper DC limit is 30V.
I'm guessing that's what components are rated for so using 28V power supply sounds reasonable to leave some room for overshoot.

Nevertheless, that's a noticeable improvement and once fully supported, will be a nice direct replacement for TS100 :D

**Comment by dimxbox** [comment]
> It was brought up before that input voltage can be 28V for DC (what MINIWARE officially confirms). I stumbled upon their comment [here](https://oscarliang.com/ts101-soldering-iron/#comment-158612) where they claim upper DC limit is 30V. I'm guessing that's what components are rated for so using 28V power supply sounds reasonable to leave some room for overshoot.
> 
> Nevertheless, that's a noticeable improvement and once fully supported, will be a nice direct replacement for TS100 :D

I use a Microtik 28V 3.4A power supply with the TS101 soldering iron - no problems.
Mikrotik Power Adapter MT96-280340-12DT
Model: DQS961-280340-3

**Comment by Ryyis** [comment]
I would like to flash my 101. When I connect it to PC it only shows 8kb total and free are 3kb. The hex file is to big. Have I overlooked something??

**Comment by Ryyis** [comment]
Thank you for the explanation. Ofc I havn't pressed A. Now it works.

**Comment by Wi-Fixit** [comment]
@Ralim Is there a way i can install and go back to the stock firmware? I'm sure IronOS is great if i actually understood what some of the features are for and how to use them properly. But for a simple person like me the stock firmware was easier to navigate, easier to toggle sleep/heat mode and it would boost to 450C which now it won't go over 429C no matter how many watts i push it to.

**Comment by Ralim** [comment]
You can always just flash the stock firmware back the same way, you just flash their firmwares .hex file (the same as you used to flash IronOS)

**Comment by baronkaz** [comment]
hi, where can i download ironos for ts101 now? previous link is expired. thanks

**Comment by StevenSeifried** [comment]
> hi, where can i download ironos for ts101 now? previous link is expired. thanks

https://github.com/Ralim/IronOS/releases
